event_content_collection!: implement parsing input into enum

DevinR528 opened the pull request.

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -0,0 +1,65 @@
+        .events
+        .iter()
+        .map(|lit| {
+            format!(
```
Why are you using `format!` here rather than later interpolating the `to_camel_case` and `to_event_content` output into a `quote!` (or `parse_quote!`) invocation?

jplatte replied:
I think I get it. With this, you turn the strings into idents. I just checked though and we can much improve this to keep spans "automatically" in the conversion if we just change `to_camel_case` and `to_event_content` (see review comment on `to_camel_case`).

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -0,0 +1,65 @@
+    for (idx, var) in input
+        .events
+        .iter()
+        .map(|lit| {
```
Why do this when constructing the iterator instead of in the loop body?

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -0,0 +1,65 @@
+                to_event_content(lit.value())
+            )
+        })
+        .map(|s| parse_str::<syn::Variant>(&s))
```
Same as previous map: Why not do this in the loop body?

jplatte commented on ruma-events-macros/src/collection_parse.rs:
```diff
@@ -0,0 +1,62 @@
+
+impl Parse for RumaCollectionInput {
+    fn parse(input: ParseStream<'_>) -> parse::Result<Self> {
+        let attrs = input.call(Attribute::parse_outer)?;
```
Parsing attributes in front of `name:`? Why?

DevinR528 replied:
Do we not need to parse them at all? I was imagining for docs.

jplatte replied:
Ah, I see. Well, let's keep it as-is for now then. This will probably disappear once there is one macro to generate all collections at once.

DevinR528 replied:
Oh right then it'll just be generated.

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -0,0 +1,65 @@
+
+/// Splits the given `event_type` string on `.` and `_` removing the `m.` then
+/// using only the event name append "EventContent".
+fn to_event_content(name: String) -> String {
```
Same as `to_camel_case`:

```suggestion
fn to_event_content(name: LitStr) -> Ident {
```

jplatte commented on ruma-events-macros/src/gen.rs:
```diff
@@ -180,7 +180,7 @@ fn populate_state_fields(content_name: Ident, fields: Vec<Field>) -> Vec<Field>
 /// Splits the given `event_type` string on `.` and `_` removing the `m.` then
 /// camel casing to give the `EventType` variant.
-fn to_camel_case(name: String) -> String {
+pub(crate) fn to_camel_case(name: String) -> String {
```
This function is only ever called with `lit_str.value()` and we really should be preserving spans for it (as we do in this file [on the master branch](https://github.com/ruma/ruma-events/blob/a3f377ec39490745de667d0916edfd7bc49a4289/ruma-events-macros/src/gen.rs#L83-L84)); why not make that mandatory by taking a `LitStr` and returning an `Ident`:

```suggestion
pub(crate) fn to_camel_case(name: LitStr) -> Ident {
```

Also, since this function is not actually used in this file on the api22 branch, let's move it to `collection.rs`.

jplatte commented on ruma-events-macros/src/collection_parse.rs:
```diff
@@ -0,0 +1,62 @@
+            {
+                events.push(lit_str);
+            } else {
+                let msg = "values of field `events` are required to be a string literal by `event_content_collection!`";
```
I think we don't need to tell the user of where this error is coming from if there's an error span that points right at it :)

```suggestion
                let msg = "values of field `events` are required to be a string literal";
```

jplatte commented on ruma-events-macros/src/collection_parse.rs:
```diff
@@ -0,0 +1,62 @@
+//! Details of parsing input for the `event_content_collection` procedural macro.
```
Maybe have this as a sub-module of `collection`, with the definition of `RumaCollectionInput` being part of `collection`?

jplatte commented on ruma-events-macros/src/collection_parse.rs:
```diff
@@ -0,0 +1,62 @@
+    Attribute, Expr, ExprLit, Ident, Lit, LitStr, Token,
+};
+
+mod kw {
```
To appease clippy:

```suggestion
/// Custom keywords for the `event_content_collection!` macro
mod kw {
```

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -1,40 +1,30 @@
-        var.attrs.push(parse_quote! { #[doc = #content_docstring] });
-        
-        variants.push(var)
+    for lit in input.events.iter() {
```
And now the `let mut variants = vec![];` + `for` loop that fills it can just become

```rust
let variants = input.events.iter().map(|event| {
    // ...
}).collect::<syn::Result<_>>()?;
```

:)

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -1,40 +1,30 @@
+    for lit in input.events.iter() {
+        let var = to_camel_case(lit);
+        let content = to_event_content(lit);
+        let mut variant: syn::Variant = parse_quote! { #var(#content) };
+
+        let content_docstring = lit;
+        variant
+            .attrs
+            .push(parse_quote! { #[doc = #content_docstring] });
```
Why do this in two separate steps / `parse_quote!` invocations?

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -61,5 +53,71 @@ fn to_event_content(name: String) -> String {
         .map(|event| format!("{}EventContent", event))
-        .unwrap()
+        .map(|id| Ident::new(&id, span))
+        .expect("must event type must contain more than 'm.'")
```
I think this becomes more readable if the `expect` happens right after `.last()`.

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -61,5 +53,71 @@ fn to_event_content(name: String) -> String {
+            input.parse::<kw::events>()?;
+            input.parse::<Token![:]>()?;
+            // an array of event names `["m.room.whatever"]`
+            let mut events = vec![];
+            for item in input.parse::<syn::ExprArray>()?.elems.into_iter() {
```
This can also be done by collecting into a `syn::Result`. I think it is more idiomatic to do that in cases like this, and it's even got the performance benefit of pre-allocating (though that's not going to matter here).

jplatte commented on ruma-events-macros/src/gen.rs:
```diff
@@ -180,12 +180,15 @@ fn populate_state_fields(content_name: Ident, fields: Vec<Field>) -> Vec<Field>
 /// Splits the given `event_type` string on `.` and `_` removing the `m.` then
 /// camel casing to give the `EventType` variant.
-pub(crate) fn to_camel_case(name: String) -> String {
+pub(crate) fn to_camel_case(name: &LitStr) -> Ident {
```
You didn't respond to or implement my suggestion to move this to `collection.rs` since it's not being used here since the removal of the event traits. I'm guessing you just missed it?

DevinR528 replied:
Sorry, yeah I didn't see it under the code.

jplatte commented:
I don't understand what your latest commit does. In particular, why are you no longer splitting on `_`?

DevinR528 commented:
Ahh right, we have to do it two steps I was thinking of the case `m.what_ever` would have given `Ever` the original way, but you are right I didn't cover the `_` case.

jplatte commented:
What do you mean with the two steps?

DevinR528 commented:
```rust
fn to_event_content(name: &LitStr) -> Ident {
    let span = name.span();
    let name = name.value();

    assert_eq!(&name[..2], "m.");

    let event = name[2..]
        .split('.')
        .last()
        .unwrap();

    let event = event
        .split('_')
        .map(|s| s.chars().next().unwrap().to_uppercase().to_string() + &s[1..])
        .collect::<String>();

    let content_str = format!("{}EventContent", event);
    Ident::new(&content_str, span)
}
```

jplatte commented:
Oh, now I get it. I could imagine re-exporting everything under the crate root with the same names as the variants we have here (so `ruma_events::room::message::MessageEventContent` becomes just `ruma_events::RoomMessageEventContent`), but that's a future API change that we shouldn't be concerned about now. The updated code looks good.

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -0,0 +1,143 @@
+            };
+            syn::parse(variant.into())
+        })
+        .collect::<syn::Result<Vec<Variant>>>()?;
```
If you put a type annotation of `Vec<_>` on the `let variants` line, you can reduce this to `collect::<syn::Result<_>>()?`, which might be more readable due to less `>`s in a row, but I don't much care either way. Feel free to ignore this.

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -0,0 +1,143 @@
+                        Err(syn::Error::new_spanned(item, msg))
+                    }
+                })
+                .collect::<syn::Result<Vec<_>>>()?;
```
Not necessary, just noticed that you can reduce this a small bit further.

```suggestion
                .collect::<syn::Result<_>>()?;
```

jplatte commented on ruma-events-macros/src/collection.rs:
```diff
@@ -0,0 +1,143 @@
+}
+
+/// Details of parsing input for the `event_content_collection` procedural macro.
+pub mod parse {
```
I'm curious (don't just change it back!): why did you now inline this module?

DevinR528 replied:
Mostly size and how related they are. I did it somewhat thinking that once `gen.rs` and `parse.rs` are gone it could be moved out again.

jplatte replied:
Okay. I'm wondering why it would be a separate module at all then. But I'm fine merging it as-is.
